sbpy/photometry/core.py

```diff
@@ -54,90 +296,374 @@ def mag(self, phase):
 
         Examples
         --------
-        >>> from sbpy.photometry import HG1G2
-        >>> hg1g2 = HG1G2(12, 0.1, -0.2) # doctest: +SKIP
-        >>> mag = hg1g2.mag([0, 5, 15, 30 ,60, 90]) # doctest: +SKIP
+        TBD
 
         not yet implemented
 
         """
 
-    def distance_module(self, eph):
-        """Account magnitudes for distance module (distance from observer,
-        distance to the Sun); return modified magnitudes
+    def mag(self, pha, **kwargs):
```

Comment:
@mkelley @migueldvb @jianyangli Guys, this is a programmatic question: this specific function requires the solar phase angle as input parameter, which can be passes as a single float, a list, or an array. My question is whether we want to keep such input parameters separate, or instead just pass an `Ephem` object that holds all these properties.

If we keep the input parameters separate, it's somewhat easy to use, unless there is a large number of them (e.g., thermal modeling routines require phase angle, r, Delta...). But even in that case, we can still pull these properties from an `Ephem` object: `pha = ephem['alpha'], ...` 

If we go with `Ephem` objects instead, people will have to generate these objects, which is easy but might be perceived as unnecessary extra work if there is only one parameter required.

What is your take on this? Go with individual parameters or DataClass objects?

Reply:
Tight integration with `Ephem` is preferred.  What I've done for the `Afrho` class is accept a dictionary-like object with 'rh', 'delta', and 'phase' keys (as needed), and assumed that `Ephem` will be compatible with this.  If only one parameter is needed, then we can accept one or the other:
```python
def mag(self, pha):
  if isinstance(pha, Ephem):
    phase = pha['phase']
  else:
    phase = pha
```
Or something similar.

Reply:
Ah, ok. It's trivial if there is only parameter required: either a quantity/list/array for that parameter or `Ephem`. I was wondering about what to do in case there are several required parameters.

So we simply assume that parameters are `DataClass` objects. If you also allow for a dictionary, you can easily turn that into a `DataClass` object with `DataClass.from_dict` inside your function.

Reply:
Well, unless there is some benefit to using `DataClass`, I think allowing any dict-like object is better for users.  `Afrho.from_fluxd(wave, flam, aper, eph={'rh': 1 * u.au, 'delta': 1 * u.au, 'phase': 60 * u.deg})` is more straightforward and less command-line typing than `Afrho.from_fluxd(wave, flam, aper, eph=Ephem.from_dict({'rh': 1 * u.au, 'delta': 1 * u.au, 'phase': 60 * u.deg}))`, not to mention the additional import statement.  Here is how I documented this parameter:
```
        eph : dictionary-like or `~Ephem`
          Ephemerides of the comet, describing heliocentric and
          geocentric distances as `~astropy.units.Quantity` via
          keywords `rh` and `delta`.  Phase angle, `phase`, is
          required if `phasecor` is enabled.
```

Reply:
I agree. What I meant was that you can do the conversion from dict to `DataClass` internally in your function, saving you code lines to distinguish between a dict and `DataClass`. Just test if the input is a dict, and turn it into a `DataClass` and then assume in the remainder of your function that you are dealing with a `DataClass` object.

Reply:
Ah, of course.

Reply:
I like the flexibility by allowing this parameter to be either array-like to pass phase angle only, or `dict` or `DataClass` to include distances.  The question is, should we do distance correction in an implicit way or an explicit way?

Personally I like an implicit way.  I.e., we perform distance correction by default.  If one distance (either solar or observer) is not provided explicitly (by either the `dict` or `DataClass` type parameter), then we simply assume it as 1 au.  What do you guys think?

Reply:
Also, I haven't had time to follow all the previous discussions about `DataClass` and study the documents, but will just assume that `DataClass.from_dict` method will take care of the conversion of `dict` to `DataClass`, include error handling.  In my code, I will simply pass the parameter to it without any extra check.  Does this sound fair?  Thanks.

Reply:
Yes and yes. I prefer the implicit distance correction - as you said, one can override it by using distances of 1 au. Also, yes, `DataClass.from_dict` works out of the box and should be very robust. It does some error handling, if you want more features we can talk about that.

I also agree that we should be somewhat flexible in the data input, but I would encourage people to use the `DataClass` objects. If they prefer dictionaries or lists, we can use `DataClass` objects internally to keep things simple. Just yesterday I started setting up a mechanism that enables alternative property names so that, e.g., `eph['r']`, eph['r_hel`'], and `eph['heldist']` provide the same data. This means that we should think about what "official" property name we want and which ones we allow as alternative names. A while ago I setup this wiki page: https://github.com/NASA-Planetary-Science/sbpy/wiki/Data-container-content. I will populate it with what I got so far tomorrow and then we should go from there.

Reply:
I think using 1 au / 1 au as a default for the class is OK.